File: apps/Lodgr.Api/Features/Contracts/Controllers/ContractsController.cs
using Lodgr.Api.Common.Pagination;
using Microsoft.AspNetCore.Mvc;

namespace Lodgr.Api.Features.Contracts;

[ApiController]
[Route("api/contracts")]
public class ContractsController(IContractService service) : ControllerBase
{
    [HttpGet]
    [ProducesResponseType(typeof(PagedResult<ContractListItemResponse>), StatusCodes.Status200OK)]
    public async Task<ActionResult<PagedResult<ContractListItemResponse>>> GetContracts(
        [FromQuery] GetContractsRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedAsync(request, cancellationToken);
        return Ok(result);
    }

    [HttpGet("{contractId:long}")]
    [ProducesResponseType(typeof(ContractDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<ContractDetailResponse>> GetContractById(
        [FromRoute] long contractId,
        CancellationToken cancellationToken)
    {
        var result = await service.GetByIdAsync(contractId, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpGet("{contractId:long}/invoices")]
    [ProducesResponseType(typeof(PagedResult<ContractInvoiceListItemResponse>), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<PagedResult<ContractInvoiceListItemResponse>>> GetContractInvoices(
        [FromRoute] long contractId,
        [FromQuery] GetContractInvoicesRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedInvoicesAsync(contractId, request, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost]
    [ProducesResponseType(typeof(ContractDetailResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<ActionResult<ContractDetailResponse>> CreateContract(
        [FromBody] CreateContractRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.CreateAsync(request, cancellationToken);
        if (result.IsSuccess && result.Value is not null)
        {
            return CreatedAtAction(nameof(GetContractById), new { contractId = result.Value.ContractId }, result.Value);
        }

        return HandleOperationError(result.Error);
    }

    [HttpPut("{contractId:long}")]
    [ProducesResponseType(typeof(ContractDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<ActionResult<ContractDetailResponse>> UpdateContract(
        [FromRoute] long contractId,
        [FromBody] UpdateContractRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.UpdateAsync(contractId, request, cancellationToken);
        if (result.IsSuccess && result.Value is not null)
        {
            return Ok(result.Value);
        }

        return HandleOperationError(result.Error);
    }

    [HttpDelete("{contractId:long}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> TerminateContract(
        [FromRoute] long contractId,
        CancellationToken cancellationToken)
    {
        var terminated = await service.TerminateAsync(contractId, cancellationToken);
        return terminated ? NoContent() : NotFound();
    }

    private ActionResult HandleOperationError(OperationError? error)
    {
        if (error is null)
        {
            return Problem(title: "Unknown error", statusCode: StatusCodes.Status500InternalServerError);
        }

        return error.Type switch
        {
            OperationErrorType.Validation => BadRequest(new { error = error.Message }),
            OperationErrorType.NotFound => NotFound(new { error = error.Message }),
            OperationErrorType.Conflict => Conflict(new { error = error.Message }),
            _ => Problem(title: error.Message, statusCode: StatusCodes.Status500InternalServerError),
        };
    }
}

File: apps/api/src/main/java/org/lodgr/api/features/contracts/ContractsController.java
package org.lodgr.api.features.contracts;

import static org.lodgr.api.features.contracts.ContractDtos.*;

import jakarta.validation.Valid;
import java.util.Map;
import org.lodgr.api.pagination.PagedResult;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Validated
@RequestMapping("/api/contracts")
public class ContractsController {
    private final ContractService contractService;

    public ContractsController(ContractService contractService) {
        this.contractService = contractService;
    }

    @GetMapping
    public ResponseEntity<PagedResult<ContractListItemResponse>> getContracts(
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) Long roomId,
            @RequestParam(required = false) Long tenantId,
            @RequestParam(required = false) org.lodgr.api.model.ContractStatus status,
            @RequestParam(required = false) org.lodgr.api.model.DepositStatus depositStatus,
            @RequestParam(required = false) String keyword) {
        GetContractsQuery query = new GetContractsQuery(pageNumber, pageSize, roomId, tenantId, status, depositStatus,
                keyword);
        return ResponseEntity.ok(contractService.getPaged(query));
    }

    @GetMapping("/{contractId}")
    public ResponseEntity<ContractDetailResponse> getContractById(@PathVariable Long contractId) {
        ContractDetailResponse result = contractService.getById(contractId);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @GetMapping("/{contractId}/invoices")
    public ResponseEntity<PagedResult<ContractInvoiceListItemResponse>> getContractInvoices(
            @PathVariable Long contractId,
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) org.lodgr.api.model.InvoiceStatus status) {
        GetContractInvoicesQuery query = new GetContractInvoicesQuery(pageNumber, pageSize, status);
        PagedResult<ContractInvoiceListItemResponse> result = contractService.getPagedInvoices(contractId, query);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @PostMapping
    public ResponseEntity<?> createContract(@Valid @RequestBody CreateContractRequest request) {
        OperationResult<ContractDetailResponse> result = contractService.create(request);
        if (result.isSuccess() && result.getValue() != null) {
            return ResponseEntity.status(HttpStatus.CREATED).body(result.getValue());
        }

        return mapOperationError(result.getError());
    }

    @PutMapping("/{contractId}")
    public ResponseEntity<?> updateContract(
            @PathVariable Long contractId,
            @Valid @RequestBody UpdateContractRequest request) {
        OperationResult<ContractDetailResponse> result = contractService.update(contractId, request);
        if (result.isSuccess() && result.getValue() != null) {
            return ResponseEntity.ok(result.getValue());
        }

        return mapOperationError(result.getError());
    }

    @DeleteMapping("/{contractId}")
    public ResponseEntity<Void> terminateContract(@PathVariable Long contractId) {
        boolean terminated = contractService.terminate(contractId);
        if (!terminated) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.noContent().build();
    }

    private ResponseEntity<?> mapOperationError(OperationResult.OperationError error) {
        if (error == null) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of("error", "Unknown error"));
        }

        return switch (error.getType()) {
            case VALIDATION -> ResponseEntity.badRequest().body(Map.of("error", error.getMessage()));
            case NOT_FOUND -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", error.getMessage()));
            case CONFLICT -> ResponseEntity.status(HttpStatus.CONFLICT).body(Map.of("error", error.getMessage()));
        };
    }
}
